From 4628e91a2956ca1efdbb061f4b63c44e3a7f4c2e Mon Sep 17 00:00:00 2001
From: "weiyong.luo" <weiyong.luo@sophgo.com>
Date: Tue, 23 Jun 2026 17:27:27 +0800
Subject: [PATCH 76/84] [bug]: fix _reg_write_mask bug for alios

[description ]: NA
[root cause  ]: NA
[jira id     ]: NA
[chip project]: cv184x
[side effects]: NA
[Affected]: NA

Change-Id: I4acb18cb13cd044e23cf6a60743723299c479ed8
(cherry picked from commit 8a56904bd87d1e015a85554e1368882947bccf89)
---
 interdrv/base/platform/alios/reg.c   | 19 +++++++++++++++----
 interdrv/vpss/chip/cv184x/vpss_hal.c |  2 +-
 2 files changed, 16 insertions(+), 5 deletions(-)

diff --git a/interdrv/base/platform/alios/reg.c b/interdrv/base/platform/alios/reg.c
index da9cd234..34696466 100644
--- a/interdrv/base/platform/alios/reg.c
+++ b/interdrv/base/platform/alios/reg.c
@@ -1,18 +1,29 @@
-#include "mmio.h"
+#include "io.h"
+#include <aos/kernel.h>
+#include <k_api.h>
+
+static kspinlock_t mask_lock;
 
 
 u32 _reg_read(uintptr_t addr)
 {
-	return mmio_read_32(addr);
+	return readl((void *)addr);
 }
 
 void _reg_write(uintptr_t addr, u32 data)
 {
-	mmio_write_32(addr, data);
+	writel(data, (void *)addr);
 }
 
 void _reg_write_mask(uintptr_t addr, u32 mask, u32 data)
 {
-	mmio_clrsetbits_32(addr, mask, data);
+	unsigned long flags;
+	u32 value;
+
+	krhino_spin_lock_irq_save(&mask_lock, flags);
+	value = readl_relaxed((void *)addr) & ~mask;
+	value |= (data & mask);
+	writel_relaxed(value, (void *)addr);
+	krhino_spin_unlock_irq_restore(&mask_lock, flags);
 }
 
diff --git a/interdrv/vpss/chip/cv184x/vpss_hal.c b/interdrv/vpss/chip/cv184x/vpss_hal.c
index c7e17ac4..25bace1d 100644
--- a/interdrv/vpss/chip/cv184x/vpss_hal.c
+++ b/interdrv/vpss/chip/cv184x/vpss_hal.c
@@ -454,7 +454,7 @@ void vpss_hal_job_finish(struct vpss_device *device)
 			return;
 		}
 
-		if (device->core_list[i]->is_sbm && !device->core_list[i]->vc_sbm_done) {
+		if (job->cfg.chn_enable[i] && device->core_list[i]->is_sbm && !device->core_list[i]->vc_sbm_done) {
 			osal_spin_unlock_irqrestore(&device->dev_lock, &flags);
 			return;
 		}
-- 
2.25.1

